pkg/process/checks/gpu_detector.go

```diff
@@ -0,0 +1,107 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package checks
+
+import (
+	"sync/atomic"
+
+	workloadmeta "github.com/DataDog/datadog-agent/comp/core/workloadmeta/def"
+	"github.com/DataDog/datadog-agent/pkg/util/common"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type GPUDetector struct {
+	detectedGPU atomic.Bool
+	gpuEventsCh chan workloadmeta.EventBundle
+	stopCh      chan struct{}
+	wmeta       workloadmeta.Component
+}
+
+func NewGPUDetector(wmeta workloadmeta.Component) *GPUDetector {
+	return &GPUDetector{
+		stopCh: make(chan struct{}),
+		wmeta:  wmeta,
+	}
+}
+
+// Run starts the GPU detector, which listens for workloadmeta events to detect GPUs on the host
+func (g *GPUDetector) Run() {
+	filter := workloadmeta.NewFilterBuilder().
+		SetSource(workloadmeta.SourceRuntime).
+		SetEventType(workloadmeta.EventTypeSet).
+		AddKind(workloadmeta.KindGPU).
+		Build()
+
+	g.gpuEventsCh = g.wmeta.Subscribe(
+		"gpu-detector-set-gpu",
+		workloadmeta.NormalPriority,
+		filter,
+	)
+	for {
+		select {
+		case eventBundle, ok := <-g.gpuEventsCh:
+			if !ok {
+				return
+			}
+			for _, event := range eventBundle.Events {
+				gpu, ok := event.Entity.(*workloadmeta.GPU)
+				if !ok {
+					log.Debugf("Expected workloadmeta.GPU got %T, skipping", event.Entity)
+					continue
+				}
+
+				log.Info("GPU detected, enabling GPU tagging:", gpu)
+				g.detectedGPU.Store(true)
+				break
+			}
+			eventBundle.Acknowledge()
+
+			if g.detectedGPU.Load() {
+				log.Info("GPU detected in event bundle")
+			} else {
+				log.Info("GPU not detected in event bundle, continuing to listen")
+			}
+		case <-g.stopCh:
+			g.wmeta.Unsubscribe(g.gpuEventsCh)
+			return
+		}
+	}
+}
+
+// GetGPUTags creates and returns a mapping of active pids to their associated GPU tags
+func (g *GPUDetector) GetGPUTags() map[int32][]string {
+	if !g.detectedGPU.Load() {
+		log.Info("GPU not detected, skipping GPU tag creation")
+		return nil
+	}
+
+	wmetaGPUs := g.wmeta.ListGPUs()
+
+	pidToTagSet := make(map[int32]common.StringSet)
+	for _, gpu := range wmetaGPUs {
+		for _, pid := range gpu.ActivePIDs {
+			if _, ok := pidToTagSet[int32(pid)]; !ok {
+				pidToTagSet[int32(pid)] = common.NewStringSet()
+			}
+			pidToTagSet[int32(pid)].Add("gpu_uuid:" + gpu.ID)
+			pidToTagSet[int32(pid)].Add("gpu_device:" + gpu.Device)
+			pidToTagSet[int32(pid)].Add("gpu_vendor:" + gpu.Vendor)
```

Comment:
Let's try not to do the tagging ourselves here. We should reuse the tagging that's already computed by the tagger, so the logic isn't scattered and to keep the tagging uniform.
See: 
This is how the tags are written to the tagger
https://github.com/DataDog/datadog-agent/blob/e8c8a77392b3f06c3ad7bd9f9ee327223687e4b6/comp/core/tagger/collectors/workloadmeta_extract.go#L623-L625

This is the nvidia check that gest the tags. I think we can do something similiar to retrieve the tags from the tagger.
https://github.com/DataDog/datadog-agent/blob/c23107ad73dbb265c858716f7a66dad6dc3d8482/pkg/collector/corechecks/gpu/nvidia/collector.go#L115